ClientResponseError from binderbot

From this GH action run: https://github.com/TomAugspurger/pangeo-dask-gateway/runs/571300961

```
2020-04-08T16:03:33.2277937Z 2020-04-08 16:03.33 Binder: Waiting on event stream (phase: launching) action=binder-start phase=event-stream
2020-04-08T16:06:57.9329437Z 2020-04-08 16:06.57 Binder: Got token and url (https://hub.binder.pangeo.io/user/tomaugspurger-p-eo-dask-gateway-3aq8of78/) action=binder-ready duration=212.04667152400003 phase=build-token
2020-04-08T16:06:57.9332526Z 2020-04-08 16:06.57 Kernel: Starting               action=kernel-start phase=start
2020-04-08T16:06:58.1466490Z 2020-04-08 16:06.58 Kernel: Started                action=kernel-start phase=complete
2020-04-08T16:06:58.1466926Z âœ… Binder and kernel started successfully.
2020-04-08T16:06:58.2965787Z /opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/site-packages/binderbot/cli.py:14: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
2020-04-08T16:06:58.2966044Z   f = asyncio.coroutine(f)
2020-04-08T16:06:58.2966176Z Traceback (most recent call last):
2020-04-08T16:06:58.2967271Z   File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/runpy.py", line 193, in _run_module_as_main
2020-04-08T16:06:58.2967469Z     return _run_code(code, main_globals, None,
2020-04-08T16:06:58.2967653Z   File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/runpy.py", line 86, in _run_code
2020-04-08T16:06:58.2968366Z     exec(code, run_globals)
2020-04-08T16:06:58.2969049Z   File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/site-packages/binderbot/cli.py", line 77, in <module>
2020-04-08T16:06:58.2969215Z     sys.exit(main())  # pragma: no cover
2020-04-08T16:06:58.2969659Z   File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/site-packages/click/core.py", line 829, in __call__
2020-04-08T16:06:58.2969824Z     return self.main(*args, **kwargs)
2020-04-08T16:06:58.2970258Z   File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/site-packages/click/core.py", line 782, in main
2020-04-08T16:06:58.2970408Z     rv = self.invoke(ctx)
2020-04-08T16:06:58.2970818Z   File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
2020-04-08T16:06:58.2970980Z     return ctx.invoke(self.callback, **ctx.params)
2020-04-08T16:06:58.2971415Z   File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/site-packages/click/core.py", line 610, in invoke
2020-04-08T16:06:58.2971580Z     return callback(*args, **kwargs)
2020-04-08T16:06:58.2972011Z   File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/site-packages/binderbot/cli.py", line 17, in wrapper
2020-04-08T16:06:58.2972160Z     return loop.run_until_complete(f(*args, **kwargs))
2020-04-08T16:06:58.2972338Z   File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
2020-04-08T16:06:58.2972487Z     return future.result()
2020-04-08T16:06:58.2972982Z   File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/site-packages/binderbot/cli.py", line 56, in main
2020-04-08T16:06:58.2973140Z     await jovyan.upload_local_notebook(fname)
2020-04-08T16:06:58.2973624Z   File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/site-packages/binderbot/binderbot.py", line 286, in upload_local_notebook
2020-04-08T16:06:58.2973789Z     await self.put_contents(notebook_filename, nb)
2020-04-08T16:06:58.2974240Z   File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/site-packages/binderbot/binderbot.py", line 150, in put_contents
2020-04-08T16:06:58.2974377Z     resp.raise_for_status()
2020-04-08T16:06:58.2974845Z   File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/site-packages/aiohttp/client_reqrep.py", line 941, in raise_for_status
2020-04-08T16:06:58.2975020Z     raise ClientResponseError(
2020-04-08T16:06:58.2975832Z aiohttp.client_exceptions.ClientResponseError: 400, message='Bad Request', url=URL('https://hub.binder.pangeo.io/user/tomaugspurger-p-eo-dask-gateway-3aq8of78/api/contents/using-dask-gateway.ipynb')
2020-04-08T16:06:58.3405683Z âŒ›ï¸ Uploading using-dask-gateway.ipynb...
2020-04-08T16:06:58.3414955Z ##[error]Process completed with exit code 1.
```

(cc @rabernat)

--- a/binderbot/binderbot.py
+++ b/binderbot/binderbot.py
@@ -16,11 +16,16 @@ import structlog
 import time
 import json
 import textwrap
+import re
 
 import nbformat
 from nbconvert.preprocessors import ClearOutputPreprocessor
 
 logger = structlog.get_logger()
+
+# https://stackoverflow.com/questions/14693701/how-can-i-remove-the-ansi-escape-sequences-from-a-string-in-python
+def _ansi_escape(text):
+    return re.compile(r'\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])').sub('', text)
 
 
 class OperationError(Exception):
@@ -35,7 +40,7 @@ class BinderUser:
         KERNEL_STARTED = 4
 
     async def __aenter__(self):
-        self.session = aiohttp.ClientSession()
+        self.session = aiohttp.ClientSession(headers={'User-Agent': 'BinderBot-cli v0.1'})
         return self
 
     async def __aexit__(self, exc_type, exc, tb):
@@ -225,8 +230,10 @@ class BinderUser:
                             #if msg_type == 'execute_result':
                             #    response = msg['content']['data']['text/plain']
                             if msg_type == 'error':
-                                traceback = msg['content']['traceback']
-                                self.log.msg('Code Execute: Error', action='code-execute', phase='error', traceback=traceback)
+                                traceback = _ansi_escape('\n'.join(msg['content']['traceback']))
+                                self.log.msg('Code Execute: Error', action='code-execute',
+                                             phase='error',
+                                             traceback=traceback)
                                 raise OperationError()
                             elif msg_type == 'stream':
                                 response = msg['content']['text']
@@ -262,10 +269,14 @@ class BinderUser:
         stdout, stderr = await self.run_code(code)
         return json.loads(stdout)
 
-    async def execute_notebook(self, notebook_filename, timeout=600):
+    async def execute_notebook(self, notebook_filename, timeout=600,
+                               env_vars={}):
+        env_var_str = str(env_vars)
         # https://nbconvert.readthedocs.io/en/latest/execute_api.html
         code = f"""
+        import os
         import nbformat
+        os.environ.update({env_var_str})
         from nbconvert.preprocessors import ExecutePreprocessor
         ep = ExecutePreprocessor(timeout={timeout})
         print("Processing {notebook_filename}")

--- a/binderbot/cli.py
+++ b/binderbot/cli.py
@@ -30,10 +30,12 @@ def coro(f):
               help="Maximum execution time (in second) for each notebook.")
 @click.option("--binder-start-timeout", default=600,
               help="Maximum time (in seconds) to wait for binder to start.")
+@click.option("--pass-env-var", "-e", multiple=True,
+              help="Environment variables to pass to the binder execution environment.")
 @click.argument('filenames', nargs=-1, type=click.Path(exists=True))
 @coro
 async def main(binder_url, repo, ref, output_dir, nb_timeout,
-               binder_start_timeout, filenames):
+               binder_start_timeout, pass_env_var, filenames):
     """Run local notebooks on a remote binder."""
 
     # validate filename inputs
@@ -49,6 +51,8 @@ async def main(binder_url, repo, ref, output_dir, nb_timeout,
                f"     repo: {repo}\n"
                f"     ref: {ref}")
 
+    extra_env_vars = {k: os.environ[k] for k in pass_env_var}
+
     # inputs look good, start up binder
     async with BinderUser(binder_url, repo, ref) as jovyan:
         await jovyan.start_binder(timeout=binder_start_timeout)
@@ -63,7 +67,8 @@ async def main(binder_url, repo, ref, output_dir, nb_timeout,
                 await jovyan.upload_local_notebook(fname)
                 click.echo("✅")
                 click.echo(f"⌛️ Executing {fname}...", nl=False)
-                await jovyan.execute_notebook(fname, timeout=nb_timeout)
+                await jovyan.execute_notebook(fname, timeout=nb_timeout,
+                                              env_vars=extra_env_vars)
                 click.echo("✅")
                 click.echo(f"⌛️ Downloading and saving {fname}...", nl=False)
                 nb_data = await jovyan.get_contents(fname)

--- a/tests/test_binderbot.py
+++ b/tests/test_binderbot.py
@@ -14,45 +14,28 @@ from binderbot import cli
 
 @pytest.fixture()
 def example_nb_data():
-    nbdata = {'cells': [{'cell_type': 'markdown',
-               'metadata': {},
-               'source': '# Test Pangeo Gallery'},
-              {'cell_type': 'code',
-               'execution_count': 1,
-               'metadata': {},
-               'outputs': [{'name': 'stdout',
-                 'output_type': 'stream',
-                 'text': "Today's date: 2020-03-12\n"}],
-               'source': 'from datetime import date\ntoday = date.today()\nprint("Today\'s date:", today)'},
-              {'cell_type': 'code',
-               'execution_count': 2,
-               'metadata': {},
-               'outputs': [{'name': 'stdout',
-                 'output_type': 'stream',
-                 'text': 'Ryans-MacBook-Pro.local\n'}],
-               'source': 'import socket\nprint(socket.gethostname())'},
-              {'cell_type': 'code',
-               'execution_count': 0,
-               'metadata': {},
-               'outputs': [],
-               'source': 'import time\ntime.sleep(2)'},
-              {'cell_type': 'code',
-               'execution_count': None,
-               'metadata': {},
-               'outputs': [],
-               'source': ''}],
-             'metadata': {'kernelspec': {'display_name': 'Python 3',
-               'language': 'python',
-               'name': 'python3'},
-              'language_info': {'codemirror_mode': {'name': 'ipython', 'version': 3},
-               'file_extension': '.py',
-               'mimetype': 'text/x-python',
-               'name': 'python',
-               'nbconvert_exporter': 'python',
-               'pygments_lexer': 'ipython3',
-               'version': '3.6.7'}},
-             'nbformat': 4,
-             'nbformat_minor': 4}
+    nbdata = {'cells': [{'cell_type': 'code',
+                       'execution_count': None,
+                       'metadata': {},
+                       'outputs': [],
+                       'source': 'import socket\nprint(socket.gethostname())'},
+                      {'cell_type': 'code',
+                       'execution_count': None,
+                       'metadata': {},
+                       'outputs': [],
+                       'source': 'import os\nprint(os.environ["MY_VAR"])'}],
+                     'metadata': {'kernelspec': {'display_name': 'Python 3',
+                       'language': 'python',
+                       'name': 'python3'},
+                      'language_info': {'codemirror_mode': {'name': 'ipython', 'version': 3},
+                       'file_extension': '.py',
+                       'mimetype': 'text/x-python',
+                       'name': 'python',
+                       'nbconvert_exporter': 'python',
+                       'pygments_lexer': 'ipython3',
+                       'version': '3.8.1'}},
+                     'nbformat': 4,
+                     'nbformat_minor': 4}
     return nbformat.from_dict(nbdata)
 
     return tmpdir
@@ -66,20 +49,23 @@ def test_cli_upload_execute_download(tmp_path, example_nb_data):
     with open(fname, 'w', encoding='utf-8') as f:
         nbformat.write(example_nb_data, f)
 
-    runner = CliRunner()
+    env = {"MY_VAR": "SECRET"}
+    runner = CliRunner(env=env)
     args = ["--binder-url", "http://mybinder.org",
             "--repo", "binder-examples/requirements",
             "--ref", "master", "--nb-timeout", "10",
+            "--pass-env-var",  "MY_VAR",
             fname]
     result = runner.invoke(cli.main, args)
     assert result.exit_code == 0, result.output
-    # assert 'binderbot.cli.main' in result.output
 
     with open(fname) as f:
         nb = nbformat.read(f, as_version=4)
 
-    hostname = nb['cells'][2]['outputs'][0]['text']
+    hostname = nb['cells'][0]['outputs'][0]['text']
     assert hostname.startswith('jupyter-binder-')
+    remote_env_var_value = nb['cells'][1]['outputs'][0]['text']
+    assert remote_env_var_value == env['MY_VAR'].rstrip()
 
     # help_result = runner.invoke(cli.main, ['--help'])
     # assert help_result.exit_code == 0

--- a/tests/test_binderbot.py
+++ b/tests/test_binderbot.py
@@ -65,7 +65,7 @@ def test_cli_upload_execute_download(tmp_path, example_nb_data):
     hostname = nb['cells'][0]['outputs'][0]['text']
     assert hostname.startswith('jupyter-binder-')
     remote_env_var_value = nb['cells'][1]['outputs'][0]['text']
-    assert remote_env_var_value == env['MY_VAR'].rstrip()
+    assert remote_env_var_value.rstrip() == env['MY_VAR']
 
     # help_result = runner.invoke(cli.main, ['--help'])
     # assert help_result.exit_code == 0

